File: Leetcode/207. Course Schedule.py
# Problem Link: https://leetcode.com/problems/course-schedule/

class Graph:
    def __init__(self, v):
        self.vertices = v
        self.adj = [[] for i in range(v)]
        
    def addEdge(self, u, v):
        self.adj[u].append(v)
        
    def dfs(self, stack, v, visited):
        visited[v] = True
        for x in self.adj[v]:
            if not visited[x]:
                self.dfs(stack, x, visited)
        stack.append(v)
        
    def topologicalSort(self):
        visited = [False for i in range(self.vertices)]
        stack = []
        for v in range(self.vertices):
            if not visited[v]:
                self.dfs(stack, v, visited)
        return stack[::-1]
    
        
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        hashTable = {}
        graph = Graph(numCourses)
        for i in range(len(prerequisites)):
            graph.addEdge(prerequisites[i][0], prerequisites[i][1])
        stack = graph.topologicalSort()
        for i in range(len(stack)):
            hashTable[stack[i]] = i
        for i in range(len(prerequisites)):
            if hashTable[prerequisites[i][0]] > hashTable[prerequisites[i][1]]:
                return False
        return True                 

File: Leetcode/207. Course Schedule.java
// Problem Link: https://leetcode.com/problems/course-schedule/
// Problem Type: Cycle Detection, Depth First Search, Recursion

class Solution {
    private List<ArrayList<Integer>> graph;
    private int[] visited;

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        graph = new ArrayList<ArrayList<Integer>>();
        for (int i = 0; i < numCourses; i++) {
            ArrayList<Integer> edge = new ArrayList<Integer>();
            graph.add(edge);
        }
        visited = new int[numCourses];
        for (int i = 0; i < prerequisites.length; i++) {
            graph.get(prerequisites[i][1]).add(prerequisites[i][0]);
        }
        for (int i = 0; i < numCourses; i++) {
            if (visited[i] == 1) continue;
            if (!dfs(i)) {
                return false;
            }
        }
        return true;
    }

    private boolean dfs(int node) {
        visited[node] = -1;
        for (Integer u : graph.get(node)) {
            if (visited[u] == 1) continue;
            if (visited[u] < 0) {
                return false;
            }
            if (!dfs(u)) return false;
        }
        visited[node] = 1;
        return true;
    }
}

/*
20
[[0,10],[3,18],[5,5],[6,11],[11,14],[13,1],[15,1],[17,4]]
output: false

3
[[1,0],[2,0],[0,2]]
output: false

2
[[1,0]]
output: true

5
[[1,4],[2,4],[3,1],[3,2]]
output: true
 */
